share/core.go

```diff
@@ -0,0 +1,315 @@
+func RecoverSecret(g abstract.Group, shares []*PriShare, t int, n int) (abstract.Scalar, error) {
+
+	isBad := func(s *PriShare) bool {
+		return s == nil || s.V == nil || s.I < 0 || n <= s.I
+	}
+
+	c := 0
+	x := make(map[int]abstract.Scalar)
+	for _, s := range shares {
+		if isBad(s) {
+			continue
+		}
+		x[s.I] = g.Scalar().SetInt64(1 + int64(s.I))
+		c++
+	}
+
+	if c < t {
+		return nil, errors.New("not enough good private shares to reconstruct shared secret")
+	}
+
+	acc := g.Scalar().Zero() // scalar sum accumulator
+	num := g.Scalar()        // numerator
+	den := g.Scalar()        // denominator
+	tmp := g.Scalar()        // scalar buffer
+
+	for _, si := range shares {
+		if isBad(si) {
+			continue
+		}
+		num.Set(si.V)
+		den.One()
+		for _, sj := range shares {
+			if isBad(sj) || sj.I == si.I {
+				continue
+			}
+			num.Mul(num, x[sj.I])
+			den.Mul(den, tmp.Sub(x[sj.I], x[si.I]))
+		}
+		acc.Add(acc, num.Div(num, den))
+	}
+
+	return acc, nil
+}
+
+// PubShare represents a public share.
+type PubShare struct {
+	I int            // Index of the public share
+	V abstract.Point // Value of the public share
+}
+
+// PubPoly represents a public commitment polynomial to a secret sharing polynomial.
+type PubPoly struct {
+	g       abstract.Group   // Cryptographic group
+	b       abstract.Point   // Base point, nil for standard base
+	commits []abstract.Point // Commitments to coefficients of the secret sharing polynomial
+}
+
+// NewPubPoly creates a new public commitment polynomial
+func NewPubPoly(g abstract.Group, b abstract.Point, commits []abstract.Point) *PubPoly {
+	return &PubPoly{g, b, commits}
+}
+
+// Info returns the base point and the commitments to the polynomial coefficients.
+func (p *PubPoly) Info() (abstract.Point, []abstract.Point) {
+	return p.b, p.commits
+}
+
+// Threshold returns the secret sharing threshold.
+func (p *PubPoly) Threshold() int {
+	return len(p.commits)
+}
+
+// GetCommit returns the secret commitment p(0), i.e., the constant term of the polynomial.
+func (p *PubPoly) GetCommit() abstract.Point {
+	return p.commits[0]
+}
+
+// Eval computes the public share v = p(i).
+func (p *PubPoly) Eval(i int) *PubShare {
+	xi := p.g.Scalar().SetInt64(1 + int64(i)) // x-coordinate of this share
+	v := p.g.Point().Null()
+	for j := p.Threshold() - 1; j >= 0; j-- {
+		v.Mul(v, xi)
+		v.Add(v, p.commits[j])
+	}
+	return &PubShare{i, v}
+}
+
+// Shares creates a list of n public commitment shares p(1),...,p(n).
+func (p *PubPoly) Shares(n int) []*PubShare {
+	shares := make([]*PubShare, n)
+	for i := 0; i < n; i++ {
+		shares[i] = p.Eval(i)
+	}
+	return shares
+}
+
+// Add computes the component-wise sum of the polynomials p and q and returns it
+// as a new polynomial. If the base points p.b and q.b are different then the
+// base point of the resulting PubPoly cannot be computed without knowing the
+// discrete logarithm between p.b and q.b. In this particular case, we are using
```

Comment:
I like to emphasize these kinds of things with a big `NOTE` or `WARNING` so it's clearly visible. what do you think ?